lib/tether/tether_linux.go

```diff
@@ -213,29 +216,104 @@ func lookPath(file string, env []string, dir string) (string, error) {
 func establishPty(session *SessionConfig) error {
 	defer trace.End(trace.Begin("initializing pty handling for session " + session.ID))
 
-	// TODO: if we want to allow raw output to the log so that subsequent tty enabled
-	// processing receives the control characters then we should be binding the PTY
-	// during attach, and using the same path we have for non-tty here
-	session.wait = &sync.WaitGroup{}
-
 	var err error
 	session.Pty, err = pty.Start(&session.Cmd)
-	if session.Pty != nil {
-		session.wait.Add(1)
+	if err != nil {
+		return err
+	}
 
-		// TODO: do we need to ensure all reads have completed before calling Wait on the process?
-		// it frees up all resources - does that mean it frees the output buffers?
-		go func() {
-			_, gerr := io.Copy(session.Outwriter, session.Pty)
-			log.Debugf("PTY stdout copy: %s", gerr)
+	session.wait.Add(1)
+	go func() {
+		_, gerr := io.CopyBuffer(session.Outwriter, session.Pty, make([]byte, ioCopyBufferSize))
+		log.Debugf("PTY stdout copy: %s", gerr)
 
-			session.wait.Done()
-		}()
+		session.wait.Done()
+	}()
+
+	go func() {
+		_, gerr := io.CopyBuffer(session.Pty, session.Reader, make([]byte, ioCopyBufferSize))
+		log.Debugf("PTY stdin copy: %s", gerr)
+	}()
+
+	return nil
+}
+
+func establishNonPty(session *SessionConfig) error {
+	defer trace.End(trace.Begin("initializing nonpty handling for session " + session.ID))
+	var err error
+
+	if session.OpenStdin {
+		log.Debugf("Setting StdinPipe")
+		if session.StdinPipe, err = session.Cmd.StdinPipe(); err != nil {
+			log.Errorf("StdinPipe failed with %s", err)
+			return err
+		}
+	}
+
+	log.Debugf("Setting StdoutPipe")
+	if session.StdoutPipe, err = session.Cmd.StdoutPipe(); err != nil {
+		log.Errorf("Setting StdoutPipe failed with %s", err)
+		return err
+	}
+
+	log.Debugf("Setting StderrPipe")
+	if session.StderrPipe, err = session.Cmd.StderrPipe(); err != nil {
+		log.Errorf("Setting StderrPipe failed with %s", err)
+		return err
+	}
+
+	if session.OpenStdin {
 		go func() {
-			_, gerr := io.Copy(session.Pty, session.Reader)
-			log.Debugf("PTY stdin copy: %s", gerr)
+			_, gerr := io.CopyBuffer(session.StdinPipe, session.Reader, make([]byte, ioCopyBufferSize))
+			log.Debugf("Reader stdin returned: %s", gerr)
+
+			if gerr == nil {
+				if cerr := session.StdinPipe.Close(); cerr != nil {
```

Comment:
that's for the other case (if the command being run will not exit until standard input is closed, the caller must close the pipe.)